File: ai-social-automation-platform/frontend/src/utils/validators.js
// Email validation
export const validateEmail = (email) => {
  if (!email) return 'Email is required'
  
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  if (!emailRegex.test(email)) {
    return 'Please enter a valid email address'
  }
  
  return null
}

// Password validation
export const validatePassword = (password, options = {}) => {
  const {
    minLength = 6,
    requireUppercase = false,
    requireLowercase = false,
    requireNumber = false,
    requireSpecial = false
  } = options
  
  if (!password) return 'Password is required'
  
  if (password.length < minLength) {
    return `Password must be at least ${minLength} characters long`
  }
  
  if (requireUppercase && !/[A-Z]/.test(password)) {
    return 'Password must contain at least one uppercase letter'
  }
  
  if (requireLowercase && !/[a-z]/.test(password)) {
    return 'Password must contain at least one lowercase letter'
  }
  
  if (requireNumber && !/\d/.test(password)) {
    return 'Password must contain at least one number'
  }
  
  if (requireSpecial && !/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
    return 'Password must contain at least one special character'
  }
  
  return null
}

// Confirm password validation
export const validateConfirmPassword = (password, confirmPassword) => {
  if (!confirmPassword) return 'Please confirm your password'
  
  if (password !== confirmPassword) {
    return 'Passwords do not match'
  }
  
  return null
}

// Name validation
export const validateName = (name, fieldName = 'Name') => {
  if (!name || !name.trim()) return `${fieldName} is required`
  
  if (name.trim().length < 2) {
    return `${fieldName} must be at least 2 characters long`
  }
  
  if (name.trim().length > 50) {
    return `${fieldName} must be less than 50 characters`
  }
  
  return null
}

// Username validation
export const validateUsername = (username) => {
  if (!username) return 'Username is required'
  
  if (username.length < 3) {
    return 'Username must be at least 3 characters long'
  }
  
  if (username.length > 30) {
    return 'Username must be less than 30 characters'
  }
  
  if (!/^[a-zA-Z0-9_]+$/.test(username)) {
    return 'Username can only contain letters, numbers, and underscores'
  }
  
  return null
}

// Phone number validation
export const validatePhoneNumber = (phone) => {
  if (!phone) return 'Phone number is required'
  
  const phoneRegex = /^[\+]?[1-9][\d]{0,15}$/
  const cleaned = phone.replace(/\D/g, '')
  
  if (!phoneRegex.test(cleaned)) {
    return 'Please enter a valid phone number'
  }
  
  return null
}

// URL validation
export const validateUrl = (url, fieldName = 'URL') => {
  if (!url) return `${fieldName} is required`
  
  try {
    new URL(url)
    return null
  } catch {
    return `Please enter a valid ${fieldName.toLowerCase()}`
  }
}

// Social media handle validation
export const validateSocialHandle = (handle, platform) => {
  if (!handle) return 'Handle is required'
  
  // Remove @ symbol if present
  const cleanHandle = handle.replace(/^@/, '')
  
  const platformRules = {
    instagram: {
      minLength: 1,
      maxLength: 30,
      pattern: /^[a-zA-Z0-9._]+$/,
      message: 'Instagram handle can only contain letters, numbers, periods, and underscores'
    },
    twitter: {
      minLength: 1,
      maxLength: 15,
      pattern: /^[a-zA-Z0-9_]+$/,
      message: 'Twitter handle can only contain letters, numbers, and underscores'
    },
    facebook: {
      minLength: 5,
      maxLength: 50,
      pattern: /^[a-zA-Z0-9.]+$/,
      message: 'Facebook username can only contain letters, numbers, and periods'
    },
    linkedin: {
      minLength: 3,
      maxLength: 100,
      pattern: /^[a-zA-Z0-9-]+$/,
      message: 'LinkedIn handle can only contain letters, numbers, and hyphens'
    },
    youtube: {
      minLength: 3,
      maxLength: 30,
      pattern: /^[a-zA-Z0-9_-]+$/,
      message: 'YouTube handle can only contain letters, numbers, underscores, and hyphens'
    }
  }
  
  const rules = platformRules[platform]
  if (!rules) return null
  
  if (cleanHandle.length < rules.minLength) {
    return `Handle must be at least ${rules.minLength} characters long`
  }
  
  if (cleanHandle.length > rules.maxLength) {
    return `Handle must be less than ${rules.maxLength} characters`
  }
  
  if (!rules.pattern.test(cleanHandle)) {
    return rules.message
  }
  
  return null
}

// Content validation
export const validateContent = (content, platform, contentType = 'post') => {
  if (!content || !content.trim()) return 'Content is required'
  
  const limits = {
    twitter: { post: 280 },
    instagram: { post: 2200, story: 2200 },
    facebook: { post: 63206 },
    linkedin: { post: 3000 },
    youtube: { description: 5000 },
    tiktok: { description: 300 }
  }
  
  const limit = limits[platform]?.[contentType] || 1000
  
  if (content.length > limit) {
    return `Content must be less than ${limit} characters for ${platform}`
  }
  
  return null
}

// File validation
export const validateFile = (file, options = {}) => {
  const {
    maxSize = 10 * 1024 * 1024, // 10MB
    allowedTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif'],
    fieldName = 'File'
  } = options
  
  if (!file) return `${fieldName} is required`
  
  if (file.size > maxSize) {
    const sizeMB = Math.round(maxSize / (1024 * 1024))
    return `${fieldName} must be smaller than ${sizeMB}MB`
  }
  
  if (!allowedTypes.includes(file.type)) {
    const types = allowedTypes.map(type => type.split('/')[1]).join(', ')
    return `${fieldName} must be one of: ${types}`
  }
  
  return null
}

// Date validation
export const validateDate = (date, fieldName = 'Date') => {
  if (!date) return `${fieldName} is required`
  
  const selectedDate = new Date(date)
  const now = new Date()
  
  if (isNaN(selectedDate.getTime())) {
    return `Please enter a valid ${fieldName.toLowerCase()}`
  }
  
  if (selectedDate < now) {
    return `${fieldName} must be in the future`
  }
  
  return null
}

// Time validation
export const validateTime = (time, fieldName = 'Time') => {
  if (!time) return `${fieldName} is required`
  
  const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/
  if (!timeRegex.test(time)) {
    return `Please enter a valid ${fieldName.toLowerCase()} (HH:MM)`
  }
  
  return null
}

// Required field validation
export const validateRequired = (value, fieldName = 'Field') => {
  if (value === null || value === undefined || value === '' || 
      (Array.isArray(value) && value.length === 0) ||
      (typeof value === 'object' && Object.keys(value).length === 0)) {
    return `${fieldName} is required`
  }
  
  return null
}

// Numeric validation
export const validateNumber = (value, options = {}) => {
  const {
    min,
    max,
    integer = false,
    fieldName = 'Value'
  } = options
  
  if (!value && value !== 0) return `${fieldName} is required`
  
  const num = parseFloat(value)
  
  if (isNaN(num)) {
    return `${fieldName} must be a valid number`
  }
  
  if (integer && !Number.isInteger(num)) {
    return `${fieldName} must be a whole number`
  }
  
  if (min !== undefined && num < min) {
    return `${fieldName} must be at least ${min}`
  }
  
  if (max !== undefined && num > max) {
    return `${fieldName} must be no more than ${max}`
  }
  
  return null
}

// Length validation
export const validateLength = (value, options = {}) => {
  const {
    min = 0,
    max = Infinity,
    fieldName = 'Field'
  } = options
  
  if (!value) return `${fieldName} is required`
  
  const length = value.length
  
  if (length < min) {
    return `${fieldName} must be at least ${min} characters long`
  }
  
  if (length > max) {
    return `${fieldName} must be no more than ${max} characters long`
  }
  
  return null
}

// Array validation
export const validateArray = (array, options = {}) => {
  const {
    minItems = 0,
    maxItems = Infinity,
    fieldName = 'Items'
  } = options
  
  if (!Array.isArray(array)) {
    return `${fieldName} must be a list`
  }
  
  if (array.length < minItems) {
    return `Please select at least ${minItems} ${fieldName.toLowerCase()}`
  }
  
  if (array.length > maxItems) {
    return `Please select no more than ${maxItems} ${fieldName.toLowerCase()}`
  }
  
  return null
}

// Form validation helper
export const validateForm = (data, rules) => {
  const errors = {}
  
  Object.keys(rules).forEach(field => {
    const fieldRules = Array.isArray(rules[field]) ? rules[field] : [rules[field]]
    const value = data[field]
    
    for (const rule of fieldRules) {
      const error = rule(value)
      if (error) {
        errors[field] = error
        break // Stop at first error for this field
      }
    }
  })
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

// Async validation helper
export const validateAsync = async (value, asyncValidator) => {
  try {
    const result = await asyncValidator(value)
    return result
  } catch (error) {
    return error.message || 'Validation failed'
  }
}

export default {
  validateEmail,
  validatePassword,
  validateConfirmPassword,
  validateName,
  validateUsername,
  validatePhoneNumber,
  validateUrl,
  validateSocialHandle,
  validateContent,
  validateFile,
  validateDate,
  validateTime,
  validateRequired,
  validateNumber,
  validateLength,
  validateArray,
  validateForm,
  validateAsync
}

File: ai-social-automation-platform/backend/app/utils/validators.py
"""
Validation utilities for user input, credentials, and content
"""
import re
from typing import Dict, List, Optional, Any
import logging

logger = logging.getLogger(__name__)

class ValidationError(Exception):
    """Custom validation exception"""
    pass

def validate_email(email: str) -> bool:
    """Validate email address format"""
    try:
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))
    except:
        return False

def validate_username(username: str) -> bool:
    """Validate username format"""
    if not username or len(username) < 3 or len(username) > 50:
        return False
    
    # Allow alphanumeric, underscore, and period
    pattern = r'^[a-zA-Z0-9._]+$'
    return bool(re.match(pattern, username))

def validate_password(password: str) -> Dict[str, Any]:
    """Validate password strength"""
    errors = []
    score = 0
    
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    else:
        score += 1
    
    if len(password) >= 12:
        score += 1
    
    if re.search(r'[a-z]', password):
        score += 1
    else:
        errors.append("Password must contain lowercase letters")
    
    if re.search(r'[A-Z]', password):
        score += 1
    else:
        errors.append("Password must contain uppercase letters")
    
    if re.search(r'\d', password):
        score += 1
    else:
        errors.append("Password must contain numbers")
    
    if re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        score += 1
    else:
        errors.append("Password must contain special characters")
    
    strength_levels = {
        0: 'very_weak',
        1: 'weak', 
        2: 'weak',
        3: 'medium',
        4: 'strong',
        5: 'very_strong',
        6: 'very_strong'
    }
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'strength': strength_levels.get(score, 'very_weak'),
        'score': score
    }

def validate_url(url: str) -> bool:
    """Validate URL format"""
    try:
        from urllib.parse import urlparse
        result = urlparse(url)
        return all([result.scheme, result.netloc])
    except:
        return False

def validate_hashtags(hashtags: List[str]) -> Dict[str, Any]:
    """Validate hashtag format"""
    errors = []
    valid_hashtags = []
    
    for hashtag in hashtags:
        # Remove # if present
        clean_tag = hashtag.lstrip('#').strip()
        
        if not clean_tag:
            errors.append("Empty hashtag found")
            continue
        
        if len(clean_tag) > 30:
            errors.append(f"Hashtag '{clean_tag}' is too long (max 30 characters)")
            continue
        
        # Check for valid characters (alphanumeric and underscore)
        if not re.match(r'^[a-zA-Z0-9_]+$', clean_tag):
            errors.append(f"Hashtag '{clean_tag}' contains invalid characters")
            continue
        
        valid_hashtags.append(f"#{clean_tag}")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'valid_hashtags': valid_hashtags
    }

def validate_content_length(content: str, platform: str) -> Dict[str, Any]:
    """Validate content length for specific platforms"""
    platform_limits = {
        'instagram': {
            'caption': 2200,
            'story_text': 500
        },
        'facebook': {
            'post': 63206,
            'comment': 8000
        },
        'youtube': {
            'title': 100,
            'description': 5000
        },
        'twitter': {
            'tweet': 280,
            'thread': 280
        },
        'linkedin': {
            'post': 3000,
            'article': 110000
        }
    }
    
    limits = platform_limits.get(platform, {})
    if not limits:
        return {'is_valid': True, 'errors': [], 'length': len(content)}
    
    errors = []
    length = len(content)
    
    # Check against the first limit (usually the main content type)
    main_limit = next(iter(limits.values()))
    
    if length > main_limit:
        errors.append(f"Content too long for {platform} (max {main_limit} characters, got {length})")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'length': length,
        'limit': main_limit
    }

def validate_credentials(platform: str, credentials: Dict[str, str]) -> Dict[str, Any]:
    """Validate credentials for specific platforms"""
    errors = []
    
    if platform == 'instagram':
        if not credentials.get('username'):
            errors.append("Instagram username is required")
        elif not validate_username(credentials['username']):
            errors.append("Invalid Instagram username format")
        
        if not credentials.get('password'):
            errors.append("Instagram password is required")
        elif len(credentials['password']) < 6:
            errors.append("Instagram password is too short")
    
    elif platform == 'facebook':
        if not credentials.get('email'):
            errors.append("Facebook email is required")
        elif not validate_email(credentials['email']):
            errors.append("Invalid Facebook email format")
        
        if not credentials.get('password'):
            errors.append("Facebook password is required")
    
    elif platform == 'youtube':
        if not credentials.get('email'):
            errors.append("YouTube email is required")
        elif not validate_email(credentials['email']):
            errors.append("Invalid YouTube email format")
        
        if not credentials.get('password'):
            errors.append("YouTube password is required")
    
    elif platform == 'twitter':
        if not credentials.get('username'):
            errors.append("Twitter username is required")
        
        if not credentials.get('password'):
            errors.append("Twitter password is required")
    
    elif platform == 'linkedin':
        if not credentials.get('email'):
            errors.append("LinkedIn email is required")
        elif not validate_email(credentials['email']):
            errors.append("Invalid LinkedIn email format")
        
        if not credentials.get('password'):
            errors.append("LinkedIn password is required")
    
    else:
        errors.append(f"Unsupported platform: {platform}")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'platform': platform
    }

def validate_content(content_data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate content data before posting"""
    errors = []
    
    # Validate required fields
    if not content_data.get('text') and not content_data.get('media_url'):
        errors.append("Content must have either text or media")
    
    # Validate text content
    if content_data.get('text'):
        text = content_data['text']
        if len(text.strip()) == 0:
            errors.append("Text content cannot be empty")
        elif len(text) > 10000:  # General limit
            errors.append("Text content is too long")
    
    # Validate platform
    platform = content_data.get('platform')
    if not platform:
        errors.append("Platform is required")
    elif platform not in ['instagram', 'facebook', 'youtube', 'twitter', 'linkedin']:
        errors.append(f"Unsupported platform: {platform}")
    
    # Platform-specific validation
    if platform and content_data.get('text'):
        platform_validation = validate_content_length(content_data['text'], platform)
        if not platform_validation['is_valid']:
            errors.extend(platform_validation['errors'])
    
    # Validate hashtags if present
    if content_data.get('hashtags'):
        hashtag_validation = validate_hashtags(content_data['hashtags'])
        if not hashtag_validation['is_valid']:
            errors.extend(hashtag_validation['errors'])
    
    # Validate media URL if present
    if content_data.get('media_url'):
        if not validate_url(content_data['media_url']):
            errors.append("Invalid media URL format")
    
    # Validate scheduled time if present
    if content_data.get('scheduled_time'):
        try:
            from datetime import datetime
            if isinstance(content_data['scheduled_time'], str):
                datetime.fromisoformat(content_data['scheduled_time'].replace('Z', '+00:00'))
        except ValueError:
            errors.append("Invalid scheduled time format")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'data': content_data
    }

def validate_user_registration(user_data: Dict[str, str]) -> Dict[str, Any]:
    """Validate user registration data"""
    errors = []
    
    # Validate email
    email = user_data.get('email', '').strip()
    if not email:
        errors.append("Email is required")
    elif not validate_email(email):
        errors.append("Invalid email format")
    
    # Validate username
    username = user_data.get('username', '').strip()
    if not username:
        errors.append("Username is required")
    elif not validate_username(username):
        errors.append("Invalid username format (3-50 chars, alphanumeric and . _ only)")
    
    # Validate password
    password = user_data.get('password', '')
    if not password:
        errors.append("Password is required")
    else:
        password_validation = validate_password(password)
        if not password_validation['is_valid']:
            errors.extend(password_validation['errors'])
    
    # Validate full name
    full_name = user_data.get('full_name', '').strip()
    if not full_name:
        errors.append("Full name is required")
    elif len(full_name) < 2 or len(full_name) > 100:
        errors.append("Full name must be 2-100 characters long")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'data': {
            'email': email,
            'username': username,
            'full_name': full_name,
            'password': password
        }
    }

def validate_domain_settings(domain_data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate domain configuration settings"""
    errors = []
    
    # Validate domain name - FIXED THE SYNTAX ERROR
    name = domain_data.get('name', '').strip()
    if not name:
        errors.append("Domain name is required")
    elif not re.match(r'^[a-z0-9_]+$', name):  # FIXED: Added closing quote and $
        errors.append("Domain name must be lowercase alphanumeric with underscores")
    
    # Validate display name
    display_name = domain_data.get('display_name', '').strip()
    if not display_name:
        errors.append("Display name is required")
    elif len(display_name) > 100:
        errors.append("Display name too long (max 100 characters)")
    
    # Validate category
    valid_categories = [
        'entertainment', 'technology', 'business', 'lifestyle',
        'education', 'news', 'sports', 'health', 'fashion'
    ]
    category = domain_data.get('category')
    if not category:
        errors.append("Category is required")
    elif category not in valid_categories:
        errors.append(f"Invalid category. Must be one of: {', '.join(valid_categories)}")
    
    # Validate keywords
    keywords = domain_data.get('keywords', [])
    if isinstance(keywords, list):
        for keyword in keywords:
            if not isinstance(keyword, str) or len(keyword.strip()) == 0:
                errors.append("Invalid keyword found")
            elif len(keyword) > 50:
                errors.append(f"Keyword '{keyword}' too long (max 50 characters)")
    
    # Validate hashtags
    hashtags = domain_data.get('hashtags', [])
    if hashtags:
        hashtag_validation = validate_hashtags(hashtags)
        if not hashtag_validation['is_valid']:
            errors.extend(hashtag_validation['errors'])
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'data': domain_data
    }

def sanitize_input(text: str, max_length: Optional[int] = None) -> str:
    """Sanitize user input text"""
    if not isinstance(text, str):
        return ""
    
    # Remove potentially harmful characters
    sanitized = re.sub(r'[<>"\']', '', text)
    
    # Normalize whitespace
    sanitized = ' '.join(sanitized.split())
    
    # Truncate if necessary
    if max_length and len(sanitized) > max_length:
        sanitized = sanitized[:max_length].strip()
    
    return sanitized

def validate_file_upload(file_data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate file upload data"""
    errors = []
    
    # Check file size
    max_size = 50 * 1024 * 1024  # 50MB
    file_size = file_data.get('size', 0)
    if file_size > max_size:
        errors.append(f"File too large (max {max_size // (1024*1024)}MB)")
    
    # Check file type
    allowed_types = {
        'image': ['jpg', 'jpeg', 'png', 'gif', 'webp'],
        'video': ['mp4', 'mov', 'avi', 'mkv', 'webm'],
        'document': ['pdf', 'doc', 'docx', 'txt']
    }
    
    filename = file_data.get('filename', '')
    if not filename:
        errors.append("Filename is required")
    else:
        extension = filename.split('.')[-1].lower()
        file_type = file_data.get('type', 'image')
        
        if file_type not in allowed_types:
            errors.append(f"Unsupported file type: {file_type}")
        elif extension not in allowed_types[file_type]:
            errors.append(f"Invalid file extension for {file_type}: {extension}")
    
    return {
        'is_valid': len(errors) == 0,
        'errors': errors,
        'data': file_data
    }